### PR TITLE
Add Azure Pipelines and SonarCloud to CI

### PR DESCRIPTION
**Pull request by michdolan**
This PR adds:

Linux, macOS, and Windows CI via Azure Pipelines:
https://dev.azure.com/imageworks/OpenColorIO/_build

SonarCloud static analysis (with gcov):
https://sonarcloud.io/dashboard?id=imageworks_OpenColorIO

Linux builds now use Docker images targeting VFX Reference Platform specs. These live in my personal DockerHub temporarily, but will be moved to the ASWF account. We can later link that to GitHub to automatically update images when changes are made. The Dockerfile setup employed here is based on similar work done by Dan Bailey for OpenVDB.

The SonarCloud build is triggered once weekly (on Saturday), and anytime a build is manually queued through Azure dev ops.

Azure jobs are defined in the root azure-pipelines.yml file, and SonarCloud is configured in the root sonar-project.properties file. Dependent Azure templates, shell scripts, and Dockerfiles all live under: share/ci/

### CODE REVIEW

**Comment by michdolan** [comment]
For performance reference:
- 6 Azure builds: 9 min 49 sec
- 6 Travis builds: 39 min 37 sec
- 2 AppVeyor builds: 25 min 26 sec

**Comment by hodoulp** [comment]
I did a first quick review but it will take me some time to review all the content. But I have one question:

> 6 Azure builds: 9 min 49 sec

Do you mean the complete `Azure Pipeline` (i.e. the three platforms including all the dockers) plus `SonarCloud` static analysis ?

**Comment by michdolan** [comment]
@hodoulp That time was for the CI done as a result of opening this PR, which excludes static analysis. The builds were in general faster than the other services (which could be influenced by the Azure pipelines having 2 threads enabled for the Linux and macOS builds), but also because Azure provides 10 concurrent jobs, whereas Travis only had 5 in this case, and AppVeyor had 1.

**Comment by michdolan** [comment]
The last Azure build with static analysis took 36 min 9 sec.

File: azure-pipelines.yml [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    vmImage: $(agentImage)
+  
+  steps:
+  - template: share/ci/templates/checkout.yml
+  - bash: share/ci/scripts/install_macos.sh 2.7.16
+    displayName: Install dependencies
+
+  - template: share/ci/templates/configure.yml
+    parameters:
+      cxxFlags: $(nixCxxFlags) -Wno-deprecated-register
```
[the remark this answers is not in the document]
Comment (hodoulp) [reply azure-pipelines.yml]: It would be better on long-term to add the flag `-Wno-deprecated-register` in the CMakeLists.txt

File: azure-pipelines.yml [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+# ------------------------------------------------------------------------------
+# TODO: Move Docker images to ASWF DockerHub account
+
+- job: Linux
+  strategy:
+    matrix:
+      Ubuntu 18.04 VFX CY2020:
+        agentImage: 'ubuntu-16.04'
+        containerImage: michdolan/ubuntu18.04:vfx2020_py27
+        cxxFlags: ''
```
Comment (hodoulp) [inline azure-pipelines.yml]: There is some confusion here: the name refers to 18, but the image is 16, and the container is 18.

File: azure-pipelines.yml [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+      Ubuntu 18.04 VFX CY2019:
+        agentImage: 'ubuntu-16.04'
+        containerImage: michdolan/ubuntu18.04:vfx2019
+        cxxFlags: ''
+        cxxCompiler: g++
+        cCompiler: gcc
+      Ubuntu 18.04 Clang 6.0:
+        agentImage: 'ubuntu-16.04'
+        containerImage: michdolan/ubuntu18.04:vfx2019
+        cxxFlags: -Wno-deprecated-register
```
Comment (hodoulp) [inline azure-pipelines.yml]: It would be better on long-term to add the flag `-Wno-deprecated-register` in the CMakeLists.txt

File: share/ci/scripts/install_openexr.sh [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+git clone https://github.com/openexr/openexr.git
+cd openexr
+
+if [ "$OPENEXR_VERSION" != "latest" ]; then
+    git checkout tags/v${OPENEXR_VERSION} -b v${OPENEXR_VERSION}
+fi
+
+mkdir build
+cd build
+cmake ../.
```
Comment (hodoulp) [inline share/ci/scripts/install_openexr.sh]: To be consistent with Boost & OIIO, cmake should have `-DCMAKE_INSTALL_PREFIX=/usr/local`.

File: share/ci/templates/build_sonar.yml [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    build-wrapper-linux-x86-64 --out-dir bw_output make clean all
+  workingDirectory: _build
+  displayName: Build OCIO with build-wrapper
+
+- template: test.yml
+
+- bash: share/ci/scripts/run_gcov.sh
+  displayName: Generate code coverage report
+
+- bash: sonar-scanner -X -Dsonar.login=$SONAR_TOKEN
```
Comment (hodoulp) [inline share/ci/templates/build_sonar.yml]: Is there a way to break on new `blocker`?

**Comment by hodoulp** [comment]
> The last Azure build with static analysis took 36 min 9 sec.

I would suggest to use it as the default CI build (i.e Travis takes 39 mins without the static analysis).

File: azure-pipelines.yml [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+  condition: in(variables['Build.Reason'], 'Manual', 'Schedule')
+  pool:
+    vmImage: 'ubuntu-16.04'
+  container: michdolan/ubuntu18.04:vfx2019
+
+  steps:
+  - template: share/ci/templates/checkout.yml
+  - template: share/ci/templates/configure.yml
+    parameters:
+      cxxFlags: $(nixCxxFlags) -g -O0 -fprofile-arcs -ftest-coverage
```
[the remark this answers is not in the document]
Comment (michdolan) [reply azure-pipelines.yml]: This was a point I meant to raise for discussion. I set the c++ std to 14 globally (though our CMake setup forces 11 for OCIO) because VFX Reference Platform 18/19/20 all specify 14. What are your thoughts on allowing OCIO to be built with a user defined standard, just defaulting to 11? OIIO does that - defaults to 11, but supports 14 or 17 being specified.

File: azure-pipelines.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (michdolan) [reply azure-pipelines.yml]: OCIO still builds with 11, yes. Only the dependencies are built with 14. I could change the CI build to 11 globally, but per my comment above, perhaps we should allow 14 to comply with VFX Ref Platform?

Clang builds fail without the "-Wno-deprecated-register" flag on Linux and macOS (warning about deprecation in C++17). Makes sense to move it to CMake. I could add that change to this PR?

File: azure-pipelines.yml [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+  strategy:
+    matrix:
+      2016 MSVC 14.16:
+        agentImage: 'vs2017-win2016'
+        buildSharedLibs: ON
+        buildDocs: ON
+      2016 MSVC 14.16 Static:
+        agentImage: 'vs2017-win2016'
+        buildSharedLibs: OFF
+        buildDocs: OFF
```
[the remark this answers is not in the document]
Comment (michdolan) [reply azure-pipelines.yml]: VS 2015 is supported by Azure, and can be added with agentImage: 'vs2015-win2012r2'. Would you suggest have 4 windows builds, or re-configuring to have three and meet these needs? We can have 10 jobs in parallel, so we have 4 more possible without slowing down CI.

File: azure-pipelines.yml [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+    inputs:
+      versionSpec: '2.7'
+    displayName: Configure Python
+
+  - template: share/ci/templates/checkout.yml
+  - template: share/ci/templates/configure.yml
+    parameters:
+      buildSharedLibs: $(buildSharedLibs)
+      buildDocs: $(buildDocs)
+      cxxFlags: /std:c++$(cxxStd) /EHsc /D WIN32 /D NDEBUG
```
[the remark this answers is not in the document]
Comment (michdolan) [reply azure-pipelines.yml]: $(cxxStd) comes from the global var set at top of azure-pipelines.yml, so this is specifying 14, yes.

Without forcing "/D WIN32", some of our code and dependencies fail, as there are some preprocessor conditionals depending on that being defined. But yes, we could move it out of here. Makes sense.

File: azure-pipelines.yml (hunk 2 above)
Comment (michdolan) [reply azure-pipelines.yml]: In this case, the 16 agentImage is just the host machine, but the build is done in the Docker container which is 18. Azure only offers Ubuntu 16 for Linux builds, but the Docker container fully insulates from that, so we can do builds in CentOS, or whatever other Linux distro we want.

File: azure-pipelines.yml (hunk 3 above)
Comment (michdolan) [reply azure-pipelines.yml]: Sounds good.

File: share/ci/docker/ubuntu18.04_vfx2020_py27/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+# TODO: Change to Python 3.7 when OCIO supports it
+
+RUN /tmp/install.sh 6 6.0 2.7 && \
+    /tmp/install_numpy.sh 1.16 && \
+    /tmp/install_cmake.sh 3 14 4 && \
+    /tmp/install_truelight.sh 4 0 7863 && \
+    /tmp/install_boost.sh 1 70 0 && \
+    /tmp/install_openexr.sh 2.3.0 && \
+    /tmp/install_oiio.sh 2.0.8 && \
+    /tmp/install_sonar.sh 3.3.0.1492
```
[the remark this answers is not in the document]
Comment (michdolan) [reply share/ci/docker/ubuntu18.04_vfx2020_py27/Dockerfile]: Good points. I will need to add another step which just cleans up apt residuals, which is fine. These were bundled together because the installs are all managed by apt, so are somewhat contained. A cleanup script is logical though, so I'll make that change.

File: share/ci/templates/build_sonar.yml (hunk 5 above)
Comment (michdolan) [reply share/ci/templates/build_sonar.yml]: I think the recommended approach from Sonar is to rely on notifications (which you can subscribe to) and status badges (which I should add to our README) to indicate a failed quality gate. In our case, since we will do analysis once weekly, even a failed CI build in this case could go unnoticed (without CI failure notifications). If we were using analysis for PRs, I think it would be possible to block a PR merge.

**Comment by michdolan** [comment]
I can address most of the feedback fairly quickly to resolve before a merge. If something seems more involved I can make note of it for a future PR.

File: azure-pipelines.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (michdolan) [reply azure-pipelines.yml]: Test is failing with SSE disabled:
FAILED: Index: 19 - Values: 1.40423405 and: 1.40423429 - Threshold: 1.00000001e-07
[335/555] [GammaOps / apply_moncurve_style_rev               ] - FAILED

[the remark this answers is not in the document]
Comment (hodoulp) [reply azure-pipelines.yml]: On which platform?

[the remark this answers is not in the document]
Comment (michdolan) [reply azure-pipelines.yml]: Linux

[the remark this answers is not in the document]
Comment (hodoulp) [reply azure-pipelines.yml]: Another strange issue. The `errorThreshold` seems not small enough to trap the difference on Linux platform (but enough on Windows). ~~To unlock you, please try to remove lines 617, 619, 620, 621, 622 & 623 in the file `src/OpenColorIO/ops/Gamma/GammaOps.cpp`.~~

That theory comes from my testings using the Ubuntu emulation on the Windows WSL layer (i.e. Windows 10).  But the difference is perfectly detected when using a `docker` from docker tools on Windows 7.

**In conclusion, I think that you could comment the 'Not SSE' image to unlock your work and to give us more time to investigate the `Ubuntu WSL` characteristics i.e. limitations (?) / bugs (?) .**

[the remark this answers is not in the document]
Comment (hodoulp) [reply azure-pipelines.yml]: I think that you could comment the 'Not SSE' image to unlock your work and to give us more time to investigate that problem.

File: azure-pipelines.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (hodoulp) [reply azure-pipelines.yml]: Agree

File: azure-pipelines.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (hodoulp) [reply azure-pipelines.yml]: Yes. Ideally on compilation flags dedicated to clang.

I dd some searches and you could use `if ("${CMAKE_CXX_COMPILER_ID}" MATCHES "Clang")` to detect clang.

File: azure-pipelines.yml (hunk 2 above)
Comment (hodoulp) [reply azure-pipelines.yml]: ok. Please add a one-line explanation for readability.

File: share/ci/templates/build_sonar.yml (hunk 5 above)
Comment (hodoulp) [reply share/ci/templates/build_sonar.yml]: ok. There is no short-term need to slowdown the pull request for this improvement.

**Comment by michdolan** [comment]
Summary of recent changes:

I commented out the VS 2015 and Linux Non-SSE builds since they were failing. TODOs are present in azure-pipelines.yml with an explanation of each needed investigation.

The root CMakeLists.txt now supports C++14 for OCIO. It defaults to C++11, but the CI builds use 14. Building successfully on all platforms.

I had to update ext/patch.cpp Windows platform detection or the IlmBase build would fail on Windows. When I was defining all the cxxFlags in the CI properties, it propagated to all ext builds, but not when defined in CMake.

I also added a macOS 10.13 build, since it's available from Azure.

**Comment by hodoulp** [comment]
@michdolan Thanks for amazing work you do around the Azure pipeline (i.e. CI builds & static analysis). We already have concrete feedback and pull requests which is great.

While implementing the Azure pipeline you faced various issues around compilation and unit tests which require much more investigations. I tried to help you live without finding the right answers/fixes. These points need more investigations as they could highlight deeper problematic impacting reliability of the CI builds. 

During the next days I will try to put some time in order to have answers and fixes when needed.

For short-term, that's a go.

**Comment by hodoulp** [comment]
[new_compiler_changes.txt](https://github.com/imageworks/OpenColorIO/files/3263651/new_compiler_changes.txt)

The file contains some changes:

1. Enable the `Not SSE` build
2. Remove the compilation flag `/EHsc` for MSVC
3. Plus few other warning removals

**Comment by michdolan** [comment]
Thanks @hodoulp ! I'll get these changes integrated.

**Comment by michdolan** [comment]
With your changes @hodoulp the "No SSE" builds does pass all tests. Thanks!

I attempted to remove /EHsc, but got new test failures as a result, and many more warningings during build, so I reverted just that change. See these logs for specific errors: https://dev.azure.com/imageworks/OpenColorIO/_build/results?buildId=175

I created a new Windows dependency install powershell script to get a compatible version of Python and CMake on the VS2015 host, so have added that into the matrix now too. I'm going to see if I can port that over to bash for consistency with other builds later.

File: share/ci/docker/ubuntu18.04_vfx2019/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+RUN /tmp/linux/apt/install.sh && \
+    /tmp/linux/apt/install_gcc.sh 6 && \
+    /tmp/linux/apt/install_clang.sh 6.0 && \
+    /tmp/linux/apt/install_python.sh 2.7 && \
+    /tmp/linux/install_numpy.sh 1.14 && \
+    /tmp/linux/install_cmake.sh 3.14.4 && \
+    /tmp/linux/install_truelight.sh 4.0.7863 && \
+    /tmp/linux/install_boost.sh 1.66.0 && \
+    /tmp/linux/install_openexr.sh 2.3.0 && \
+    /tmp/linux/install_oiio.sh 2.0.8 && \
```
Comment (lgritz) [inline share/ci/docker/ubuntu18.04_vfx2019/Dockerfile]: FYI, OIIO has a "release" branch marker that is kept up to whatever is the latest official production release. Using that instead of a hard-coded version might save you the trouble of having to bump this number every month or so that we do OIIO patch releases.

**Comment by michdolan** [comment]
ASWF VFX Reference Platform Docker work:

https://github.com/AcademySoftwareFoundation/aswf-docker/pull/3

**Comment by michdolan** [comment]
I tried enabling static analysis in PR builds, but am hitting SonarCloud authentication issues in that context. I *think* it will be resolved after the new CI system is merged, since it seems like a branch issue.

Docker images are now coming from the official ASWF repo. I also added a new build to the Linux matrix that pulls the latest OpenEXR and OIIO at build time. Once ext/ is cleaned up, we can pull the latest from other upstream dependencies there too to help catch issues with external updates.

**Comment by michdolan** [comment]
This could be merged next week if there's no further feedback. Friendly reminder to take a look if you haven't but would like to.

File: share/ci/scripts/linux/run_gcov.sh [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+mkdir _coverage
+cd _coverage
+
+# The sed command below converts from:
+#   ../_build/src/OpenColorIO/CMakeFiles/OpenColorIO.dir/ops/Exponent.gcno
+# to:
+#   ../src/OpenColorIO/ops/Exponent.cpp
+
+for g in $(find ../_build -name "*.gcno" -type f); do
+    gcov -p -o $(dirname "$g") $(echo "$g" | sed -e 's/\/_build\//\//' -e 's/\.gcno/\.cpp/' -e 's/\/CMakeFiles.*\.dir\//\//')
```
Comment (danrbailey) [inline share/ci/scripts/linux/run_gcov.sh]: @michdolan - I've been experimenting with SonarCloud recently and copied some of what you're doing here (thanks!). One issue I ran into is that the code coverage for header files didn't seem correct. I believe the issue was because many of our headers were being #included in more than one translation unit, so the gcov files for the headers were being overwritten.

Long story short, including the --long-file-names (-l) flag to gcov solved the problem because it prefixed the .cpp filename, hence making the header gcov files unique. See:

https://gcc.gnu.org/onlinedocs/gcc/Invoking-Gcov.html

Mentioning this in case it's useful here (though I'll point out that I've been mainly using clang + llvm-cov gcov so might not be relevant).

Comment (michdolan) [reply share/ci/scripts/linux/run_gcov.sh]: Thanks @danrbailey ! Good catch. I'll give that a try.
